src/datumaro/components/environment.py

```diff
@@ -8,127 +8,39 @@
-        self._filter = filter
-
-    def get(self, key: str) -> PLUGIN_TYPES:
-        """Returns a class or a factory function"""
-        item = self._items[key]
-        if issubclass(item, LazyPlugin):
-            return item.get_plugin_cls()
-        return item
-
-    def batch_register(self, values: Iterable[CliPlugin]):
-        for v in values:
-            if self._filter and not self._filter(v):
-                continue
-
-            self.register(v.NAME, v)
+from datumaro.components.registry import (
+    DatasetBaseRegistry,
+    ExporterRegistry,
+    GeneratorRegistry,
+    ImporterRegistry,
+    LauncherRegistry,
+    PluginRegistry,
+    TransformRegistry,
+    ValidatorRegistry,
+)
+from datumaro.util.os_util import get_all_file_extensions, import_foreign_module, split_path
 
 
 class Environment:
     _builtin_plugins = None
 
-    @classmethod
-    def _make_filter(cls, accept, decline=None, skip=None):
-        accept = (accept,) if isclass(accept) else tuple(accept)
-        skip = {skip} if isclass(skip) else set(skip or [])
-        skip = tuple(skip | set(accept))
-        return partial(cls._check_type, accept=accept, decline=decline, skip=skip)
-
-    @staticmethod
-    def _check_type(t, *, accept, decline, skip):
-        if not issubclass(t, accept) or t in skip or (decline and issubclass(t, decline)):
-            return False
-        if getattr(t, "__not_plugin__", None):
-            return False
-        return True
-
     def __init__(self, use_lazy_import: bool = True):
-        from datumaro.components.dataset_base import DatasetBase, SubsetBase
-        from datumaro.components.exporter import Exporter
-        from datumaro.components.generator import DatasetGenerator
-        from datumaro.components.importer import Importer
-        from datumaro.components.launcher import Launcher
-        from datumaro.components.transformer import ItemTransform, Transform
-        from datumaro.components.validator import Validator
-
-        _filter = self._make_filter
-        self._extractors = PluginRegistry(
-            _filter(
-                DatasetBase,
-                decline=Transform,
-                skip=(SubsetBase, Transform, ItemTransform),
-            )
-        )
-        self._importers = PluginRegistry(_filter(Importer))
-        self._launchers = PluginRegistry(_filter(Launcher))
-        self._exporters = PluginRegistry(_filter(Exporter))
-        self._generators = PluginRegistry(_filter(DatasetGenerator))
-        self._transforms = PluginRegistry(_filter(Transform, skip=ItemTransform))
-        self._validators = PluginRegistry(_filter(Validator))
+        self._extractors = DatasetBaseRegistry()
+        self._importers = ImporterRegistry()
+        self._launchers = LauncherRegistry()
+        self._exporters = ExporterRegistry()
+        self._generators = GeneratorRegistry()
+        self._transforms = TransformRegistry()
+        self._validators = ValidatorRegistry()
```

Comment:
Can I ask you about the benefit by separating registries as above? Is there any number (time analysis) with better representations?

Reply:
This reorganization does not reduce execution time but offers significant benefits:

- It enables `ImporterRegistry` to implement its customized `register` method, populating the `extension_groups` property efficiently. This empowers all registries to override core methods from the base `PluginRegistry` as required.
- It relocates the type definitions for stored, skipped, and ignored `CliPlugin` subclasses within the respective `PluginRegistry` implementations. This enhances clarity and organization within the codebase (IMHO).